<a href="https://colab.research.google.com/github/jnramirezg/mecanica_de_solidos/blob/main/codigo/23-(4_8_3)-ejemplo_DP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 23. Relación entre los esfuerzos principales obtenidos en el análisis bidimensional y tridimensional (sección 4.8.3)

|Quién | Fecha | Qué hizo |
| ---  | ---   | ---      |
|Michael Heredia Pérez, <mherediap@unal.edu.co> | 2022-07-20 | Conversión de código de MATLAB a Python|

## Ejemplo de *deformación plana*


Al igual que como hicimos con sympy, podemos importar completamente la librería de numpy para trabajar con ella, sin embargo, recuerden que no es la práctica recomendada.

In [1]:
from numpy import *
set_printoptions(precision=4)   # imprime todo con 4 cifras decimales max

Asumamos $\nu = $0.30 y un estado de **deformación plana**; para calcular los nuevos esfuerzos y direcciones principales, adicionalmente a $\sigma_x = $ − 1 Pa, $\sigma_y = $ = 2 Pa y $\tau_{xy} = $ − 3 Pa, se debe tener en cuenta que $\tau_{xz} = G\gamma_{xz}$ = 0 Pa, $\tau_{yz} = G\gamma_{yz}$ = 0 Pa y $\sigma_z = \nu(\sigma_x + \sigma_v) = $ 0.3( − 1 Pa + 2 Pa) = 0.3 Pa (este último esfuerzo se calculó utilizando la ecuación (4.42)). Para calcular los valores y vectores propios del problema ensamblamos la matriz (4.43)

In [2]:
# Defino las variables.
sx  = -1
sy  = 2
txy = -3
nu  = 0.30

# La matriz de esfuerzos.
sigma = array([[sx, txy,            0],
               [txy, sy,            0],
               [0,    0, nu*(sx + sy)]])
sigma

array([[-1. , -3. ,  0. ],
       [-3. ,  2. ,  0. ],
       [ 0. ,  0. ,  0.3]])

In [3]:
# Calculo lo vectores y valores propios
valp, vecp = linalg.eig(sigma)

In [4]:
valp

array([-2.8541,  3.8541,  0.3   ])

In [5]:
vecp

array([[-0.8507,  0.5257,  0.    ],
       [-0.5257, -0.8507,  0.    ],
       [ 0.    ,  0.    ,  1.    ]])

Es necesario enumerar de nuevo los valores y vectores propios de modo que $\sigma_1 \geq \sigma_2 \geq \sigma_3$, y organizar el sentido de los vectores propios para que se cumpla la regla de la mano derecha ($\mathbf{ \hat{n}_3 = \hat{n}_1 \times \hat{n}_2 }$), por lo que la solución al ejercicio se debe dar como:

$\sigma_3 , \sigma_2 , \sigma_1 =$

In [6]:
sort(valp)

array([-2.8541,  0.3   ,  3.8541])

$\hat{n}_1 = $

In [7]:
vecp[:,1]

array([ 0.5257, -0.8507,  0.    ])

$\hat{n}_2 = $

In [8]:
vecp[:,2]

array([0., 0., 1.])

$\hat{n}_3 = $

In [9]:
vecp[:,0]

array([-0.8507, -0.5257,  0.    ])

ya que al hacer el producto cruz, obtenemos

In [10]:
n1 = vecp[:, 1]
n2 = vecp[:, 2]
cross(n1, n2)

array([-0.8507, -0.5257,  0.    ])

con lo cual cumplimos la regla de la mano derecha, al obtener el sentido requerido para $\hat{n}_3$.

Si se hubiera asumido el vector $\hat{n}_2 = $ como $[0, 0, 1]^\intercal$, el sentido del vector $\hat{n}_3$ sería el dado por el siguiente cálculo:

In [11]:
n1 = vecp[:, 1]
n2 = -vecp[:, 2]
cross(n1, n2)

array([ 0.8507,  0.5257, -0.    ])

Observe que en este caso, dos de los valores propios coincidieron con $(\sigma_1)_{xy} =$ 3.8541 Pa y $(\sigma_2)_{xy} =$ − 2.8541 Pa, mientras que el otro valor propio es igual al $\sigma_z$ de
la ecuación (4.42), es decir, $\sigma_ = \nu (\sigma_x + \sigma_y)$ . De hecho, las direcciones principales asociadas a tales valores propios describieron los vectores $[ − 0.5257, 0.8507, 0]^\intercal$ y
$[0.8507, 0.5257, 0]^\intercal$ que habíamos mencionado anteriormente y el vector propio asociado al valor propio de $\sigma_z$ fue $[0, 0, − 1]^\intercal$ (ver Figura 4.17).

Finalmente, debemos recalcular el esfuerzo cortante máximo. Según se dijo en la Sección 2.9.7, este tiene una magnitud dada por la ecuación (2.75): $\tau_{max} = \frac{\sigma_1-\sigma_3}{2}$ y se presenta en los planos ortogonales a los vectores $\frac{ \hat{n}_1 - \hat{n}_3}{||\hat{n}_1 - \hat{n}_3||}$ y $\frac{ \hat{n}_1 + \hat{n}_3}{||\hat{n}_1 + \hat{n}_3||}$; por lo tanto, el esfuerzo cortante máximo es 3.3541 Pa y se presenta en los planos con normal $\frac{ \hat{n}_1 - \hat{n}_3}{||\hat{n}_1 - \hat{n}_3||} = $ $[0.2298, 0.9732, 0]^\intercal$ y
$\frac{ \hat{n}_1 + \hat{n}_3}{||\hat{n}_1 + \hat{n}_3||} =$  $[ − 0.9732, 0.2298, 0]^\intercal$.

En resumen, se tiene que dos de los valores propios corresponden a $(\sigma_1)_{xy}$ y $(\sigma_2)_{xy}$, mientras que el otro corresponde a $\sigma_z$. Los vectores propios asociados son los mismos calculados en el caso bidimensional para los dos primeros, agregando un 0 en la componente $z$, mientras que el vector propio asociado a $\sigma_z$ tiene la dirección de $[0,0, \pm 1]^\intercal$ , siendo necesario ajustar su sentido. 

De este modo, luego de reordenar los valores propios, tenemos:

$$\sigma_1 = \max\left( (\sigma_1)_{xy}, (\sigma_1)_{xy}, \nu(\sigma_x + \sigma_y) \right) $$

$$\sigma_2 = mediana\left( (\sigma_1)_{xy}, (\sigma_1)_{xy}, \nu(\sigma_x + \sigma_y) \right) $$

$$\sigma_3 = \min\left( (\sigma_1)_{xy}, (\sigma_1)_{xy}, \nu(\sigma_x + \sigma_y) \right) $$

$$ \tau_{\max} = \max\left( \frac{(\sigma_1)_{xy} - \nu(\sigma_x + \sigma_y)}{2},  \frac{(\sigma_2)_{xy} - \nu(\sigma_x + \sigma_y)}{2},  \frac{ (\sigma_1)_{xy} - (\sigma_2)_{xy} }{2}   \right)  $$ 